### PR TITLE
Assert output size of max_pooling_2d in GPU cases

### PR DESCRIPTION
**Pull request by soramichi**
In the CPU code of `max_pooling_2d`, the size of the expected output is checked before calculation (in [`chainer/utils/conv.py`](https://github.com/pfnet/chainer/blob/master/chainer/utils/conv.py#L25-L28])) so that the output never becomes a 0-dimentional tensor.

However, this check is missing in the GPU code (with and without cudnn) and thus passing an invalid input to `max_pooling_2d.py` yields a very un-friendly error as exampled below.

This PR adds assertions that check the output size of `max_pooling_2d` for GPU cases, and also improves the tests to check the added behavior.

Sample output before this PR is applied:
```
>>> import chainer
>>> x_data = chainer.cuda.cupy.random.rand(4, 4, 1, 4)
>>> x = chainer.Variable(x_data)
>>> chainer.functions.max_pooling_2d(x, 3, stride=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/soramichi/src/anaconda3/lib/python3.6/site-packages/chainer/functions/pooling/max_pooling_2d.py", line 175, in max_pooling_2d
    return MaxPooling2D(ksize, stride, pad, cover_all, use_cudnn)(x)
  File "/home/soramichi/src/anaconda3/lib/python3.6/site-packages/chainer/function.py", line 199, in __call__
    outputs = self.forward(in_data)
  File "/home/soramichi/src/anaconda3/lib/python3.6/site-packages/chainer/function.py", line 310, in forward
    return self.forward_gpu(inputs)
  File "/home/soramichi/src/anaconda3/lib/python3.6/site-packages/chainer/functions/pooling/max_pooling_2d.py", line 32, in forward_gpu
    return super(MaxPooling2D, self).forward_gpu(x)
  File "/home/soramichi/src/anaconda3/lib/python3.6/site-packages/chainer/functions/pooling/pooling_2d.py", line 65, in forward_gpu
    y_desc = cudnn.create_tensor_descriptor(y)
  File "/home/soramichi/src/anaconda3/lib/python3.6/site-packages/cupy/cudnn.py", line 78, in create_tensor_descriptor
    cudnn.setTensor4dDescriptor(desc.value, format, data_type, *arr.shape)
  File "cupy/cuda/cudnn.pyx", line 444, in cupy.cuda.cudnn.setTensor4dDescriptor (cupy/cuda/cudnn.cpp:2967)
  File "cupy/cuda/cudnn.pyx", line 449, in cupy.cuda.cudnn.setTensor4dDescriptor (cupy/cuda/cudnn.cpp:2833)
  File "cupy/cuda/cudnn.pyx", line 392, in cupy.cuda.cudnn.check_status (cupy/cuda/cudnn.cpp:2017)
cupy.cuda.cudnn.CuDNNError: CUDNN_STATUS_BAD_PARAM: b'CUDNN_STATUS_BAD_PARAM'
```

### CODE REVIEW

File: tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py [hunk 1]
```diff
@@ -75,6 +87,34 @@ def test_forward_gpu(self):
     def test_forward_gpu_no_cudnn(self):
         self.check_forward(cuda.to_gpu(self.x), False)
 
+    @attr.gpu
+    @condition.retry(3)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py]: This statement is redundant.
Please remove this.

File: tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py [hunk 2]
```diff
@@ -75,6 +87,34 @@ def test_forward_gpu(self):
...
+            x = chainer.Variable(x_data)
+            functions.max_pooling_2d(x, 3, stride=2, use_cudnn=False)
+        with self.assertRaisesRegexp(
+                AssertionError, 'Width in the output should be positive.'):
+            x_data = cuda.cupy.random.rand(4, 4, 4, 1).astype(self.dtype)
+            x = chainer.Variable(x_data)
+            functions.max_pooling_2d(x, 3, stride=2, use_cudnn=False)
+
+    @attr.cudnn
+    @condition.retry(3)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py]: This statement is redundant.
Please remove this.

**Comment by soramichi** [comment]
Thank you for the review.

I removed the decorators as suggested, but I don't really know why they are not required. Is it because the test cases do not use `to_gpu`??

**Comment by soramichi** [comment]
Oh sorry, I guess you meant `@condition.retry(3)` is redundant because the added tests are deterministic, but your comment was not about `@gpu` and `@cudnn`.

I re-introduced `@gpu` and `@cudnn`, and now only `@condition.retry(3)` is removed.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**